templates/cluster-template-high-availabilty.yaml

```diff
@@ -0,0 +1,201 @@
+          cloud-config: /etc/kubernetes/cloud.conf
+    clusterConfiguration:
+      controlPlaneEndpoint: "${OPENSTACK_CONTROLPLANE_IP}:6443"
+      imageRepository: k8s.gcr.io
+      apiServer:
+        extraArgs:
+          cloud-provider: openstack
+          cloud-config: /etc/kubernetes/cloud.conf
+        extraVolumes:
+        - name: cloud
+          hostPath: /etc/kubernetes/cloud.conf
+          mountPath: /etc/kubernetes/cloud.conf
+          readOnly: true
+      controllerManager:
+        extraArgs:
+          cloud-provider: openstack
+          cloud-config: /etc/kubernetes/cloud.conf
+        extraVolumes:
+        - name: cloud
+          hostPath: /etc/kubernetes/cloud.conf
+          mountPath: /etc/kubernetes/cloud.conf
+          readOnly: true
+        - name: cacerts
+          hostPath: /etc/certs/cacert
+          mountPath: /etc/certs/cacert
+          readOnly: true
+    joinConfiguration:
+      nodeRegistration:
+        name: '{{ ds.meta_data.name }}'
+        kubeletExtraArgs:
+          cloud-config: /etc/kubernetes/cloud.conf
+          cloud-provider: openstack
+    files:
+    - path: /etc/kubernetes/cloud.conf
+      owner: root
+      permissions: "0600"
+      content: ${OPENSTACK_CLOUD_PROVIDER_CONF_B64}
+      encoding: base64
+    - path: /etc/certs/cacert
+      owner: root
+      permissions: "0600"
+      content: ${OPENSTACK_CLOUD_CACERT_B64}
+      encoding: base64
+    ntp:
+      servers: []
+    users:
+    - name: capo
+      sudo: "ALL=(ALL) NOPASSWD:ALL"
+      sshAuthorizedKeys:
+      - "${SSH_AUTHORIZED_KEY}"
+  version: "${KUBERNETES_VERSION}"
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: OpenStackMachineTemplate
+metadata:
+  name: ${CLUSTER_NAME}-control-plane
+spec:
+  template:
+    spec:
+      flavor: ${OPENSTACK_CONTROL_PLANE_MACHINE_FLAVOR}
+      image: ${OPENSTACK_IMAGE_NAME}
+      availabilityZone: ${OPENSTACK_AVAILABILITY_ZONE}
+      cloudName: ${OPENSTACK_CLOUD}
+      cloudsSecret:
+        name: ${CLUSTER_NAME}-cloud-config
+        namespace: ${CLUSTER_NAME}
+---
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: MachineDeployment
+metadata:
+  name: "${CLUSTER_NAME}-md-0"
+spec:
+  clusterName: "${CLUSTER_NAME}"
+  replicas: ${WORKER_MACHINE_COUNT}
+  selector:
+    matchLabels:
+  template:
+    spec:
+      clusterName: "${CLUSTER_NAME}"
+      version: "${KUBERNETES_VERSION}"
+      bootstrap:
+        configRef:
+          name: "${CLUSTER_NAME}-md-0"
+          apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
+          kind: KubeadmConfigTemplate
+      infrastructureRef:
+        name: "${CLUSTER_NAME}-md-0"
+        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+        kind: OpenStackMachineTemplate
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: OpenStackMachineTemplate
+metadata:
+  name: ${CLUSTER_NAME}-md-0
+spec:
+  template:
+    spec:
+      availabilityZone: ${OPENSTACK_AVAILABILITY_ZONE}
```

Comment:
Might want to consider changing this to `failureDomain` and having the OpenStackCluster reconciler populate failure domain information in order to take advantage of the optional failureDomain support in core Cluster API.

Assuming that availabilityZone existed in v1alpha2, you could handle migration of the value to the Machine resource in a similar way that cluster-api-provider-aws is doing it (migrating v1alpha2 availabilityZone to v1alpha3 failureDomain), with a plan of removing the availabilityZone/failureDomain field from the OpenstackMachine* resources with the next breaking release.

Without support for the core Cluster API failureDomain support, the KubeadmControlPlane will not properly spread control plane Machines across different availability zones.

Reply:
I'll take a look at it later. Thx :) Currently we're not really supporting the v1alpha2 CRD with the v1alpha3 controller but implementing AZs via failure domains sounds better

Reply:
I added an implementation for failureDomains. Can you take a look if I did it right?

My understanding is that for the control plane the KubeadmControlPlane controller chooses a failureDomain based on the status of the Cluster resource.
For all other Machines (e.g. MachineDeployment) the failureDomain can (and has to be) defined in the MachineSpec